cmd/misc/main.go

```diff
@@ -1628,68 +1628,134 @@ func reIndexBlocks(start uint64, end uint64, bt *db.Bigtable, client *rpc.Erigon
+			utils.LogError(err, "error importing ens from events", 0)
+			return
 		}
-		to := from + targetCount - 1
+	}
 
-		errFields["from"] = from
-		errFields["to"] = to
-		errFields["targetCount"] = targetCount
+	cache := freecache.NewCache(100 * 1024 * 1024) // 100 MB limit
+	g := errgroup.Group{}
+	g.SetLimit(int(concurrency))
 
-		for block := from; block <= to; block++ {
-			bc, _, err := client.GetBlock(int64(block), "parity/geth")
+	for i := start; i <= end; i++ {
+		height := int64(i)
+		g.Go(func() error {
+			block, _, err := client.GetBlock(height, "geth")
 			if err != nil {
-				utils.LogError(err, fmt.Sprintf("error getting block %v from the node", block), 0, errFields)
-				return
+				return fmt.Errorf("error getting block %v from the node: %w", height, err)
 			}
-			if err := bt.SaveBlock(bc); err != nil {
-				utils.LogError(err, fmt.Sprintf("error saving block: %v ", block), 0, errFields)
-				return
+			if err := bt.SaveBlock(block); err != nil {
+				return fmt.Errorf("error saving block %v: %w", height, err)
 			}
-		}
-		indexOldEth1Blocks(from, to, batchSize, 1, transformerFlag, bt, client)
+			indexOldEth1Block(block, transformers, bt, cache)
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		panic(err)
 	}
 }
 
-func indexOldEth1Blocks(startBlock uint64, endBlock uint64, batchSize uint64, concurrency uint64, transformerFlag string, bt *db.Bigtable, client *rpc.ErigonClient) {
-	if endBlock > 0 && endBlock < startBlock {
-		utils.LogError(nil, fmt.Sprintf("endBlock [%v] < startBlock [%v]", endBlock, startBlock), 0)
-		return
+func reIndexBlocks(start uint64, end uint64, bt *db.Bigtable, client *rpc.ErigonClient, transformerFlag string, batchSize uint64, concurrency uint64) {
+	if end == math.MaxInt64 {
+		lastBlockFromBlocksTable, err := bt.GetLastBlockInBlocksTable()
+		if err != nil {
+			logrus.Errorf("error retrieving last blocks from blocks table: %v", err)
+			return
+		}
+		end = uint64(lastBlockFromBlocksTable)
 	}
-	if concurrency == 0 {
-		utils.LogError(nil, "concurrency must be greater than 0", 0)
+	transformers, importENSChanges, err := getTransformers(transformerFlag, bt)
+	if err != nil {
+		utils.LogError(nil, err, 0)
 		return
 	}
-	if bt == nil {
-		utils.LogError(nil, "no bigtable provided", 0)
-		return
+	if importENSChanges {
+		if err := bt.ImportEnsUpdates(client.GetNativeClient(), math.MaxInt64); err != nil {
+			utils.LogError(err, "error importing ens from events", 0)
+			return
+		}
 	}
 
-	transforms := make([]func(blk *types.Eth1Block, cache *freecache.Cache) (*types.BulkMutations, *types.BulkMutations, error), 0)
+	readGroup := errgroup.Group{}
+	readGroup.SetLimit(int(concurrency))
+
+	writeGroup := errgroup.Group{}
+	writeGroup.SetLimit(int(concurrency*concurrency) + 1)
+
+	cache := freecache.NewCache(100 * 1024 * 1024) // 100 MB limit
+	quit := make(chan any)
+
+	sink := make(chan *types.Eth1Block)
+	writeGroup.Go(func() error {
+		for {
+			select {
+			case block := <-sink:
```

Comment:
```suggestion
			case block, ok := <-sink:
			if !ok {
				return nil 
			}
```

I think we can add check here to handle the case where the channel might be closed